docs/rfc/block-registration.md

```diff
@@ -0,0 +1,430 @@
+
+A demonstration for how block registration could be made filterable in PHP was explored in [#5802](https://github.com/WordPress/gutenberg/pull/5802). The purpose here was to explore how plugins could have better control over the registration.
+
+Another exploration in [#5652](https://github.com/WordPress/gutenberg/pull/5652) considered using JSON as a file format to share block type definitions between JavaScript and PHP. 
+
+### Conclusions
+
+* The current approaches to client-side block type registration cannot support the proposed requirement to have all block types known outside the browser context.
+* Using a statically-defined, JSON-formatted block type definition enables easy integration in both JavaScript and PHP runtimes.
+* Registering a block type in PHP would allow for attribute default values to be assigned as dynamically generated from some external state (e.g. a database value, or localized string).
+* By default, JSON does not support localization or dynamic values.
+* On the server, a block type `icon` property can only be assigned as a string and thus cannot support SVGs and component-based icons.
+
+---
+
+## Introduction
+
+Blocks are the fundamental elements of the editor. They are the primary way in which plugins and themes can register their own functionality and extend the capabilities of the editor.
+
+## Registering a block type
+
+To register a new block type, start by creating a `block.json` file. This file:
+
+* Gives a name to the block type.
+* Defines some important metadata about the registered block type (title, category, icon, description, keywords).
+* Defines the attributes of the block type.
+* Links to the editor implementation of your block type, the save function and any other context-aware property.
+
+**Example:**
+
+
+```json
+{
+	"name": "my-plugin/notice",
+	"title": "Notice",
+	"category": "common",
+	"parent": [ "core/group" ],
+	"icon": "star",
+	"description": "Shows warning, error or success notices  ...",
+	"keywords": [ "alert", "message" ],
+	"attributes": {
+		"message": {
+			"type": "string",
+			"source": "html",
+			"selector": ".meessage"
+		}
+	},
+	"edit": "blocks/notice-edit.js",
+	"save": "blocks/notice-save.js"
+}
```

Comment:
There are also the following fields defined for some blocks:
- `merge` (e.g. Heading https://github.com/WordPress/gutenberg/blob/0fce12abc365ba85288bdb71b9115d598776fcb4/packages/block-library/src/heading/index.js#L146)